File: AStar.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <climits>
#include <string>
#include <queue>
#include <cfloat>
#include <map>

using namespace std;
using ll = long long;

const int n = 12; // jumlah gedung



struct Node {
    double x; // longitude gmaps
    double y; // latitude gmaps
};

vector<Node> coordinates = {
    /* formatnya adalah (longitude, latitude) */
    {110.37811638436851, -7.786892527493072},  // 1. agape
    {110.37825953130029, -7.786562258126049},  // 2. biblos
    {110.37842314604282, -7.78631777007192},   // 3. chara
    {110.37836850018982, -7.786012910895829},  // 4. didaktos
    {110.37811203519209, -7.785926844542299},  // 5. euodia
    {110.37804171414969, -7.786487300114521},  // 6. filia
    {110.37850078264714, -7.785753220073825},  // 7. gnosis
    {110.37836550377382, -7.785523171156505},  // 8. hagios
    {110.37855178209473, -7.7852138021969335}, // 9. iama
    {110.37854751118998, -7.784999599965238},  // 10. koinonia
    {110.37821452033853, -7.78528341456613},   // 11. logos
    {110.37854481744473, -7.785369062426726}   // 12. makarios
};

const double latitude_to_meter = 111000; // 1 latitude = 111_000 meter
const double longitude_to_meter = 72000; // 1 longitude = 72_000 meter

vector<Node> get_planar() {

    vector<Node> planar;

    double min_x = DBL_MAX;
    double min_y = DBL_MAX;
    for (auto &[longitude, latitude] : coordinates) {
        double x = longitude * longitude_to_meter; 
        double y = latitude * latitude_to_meter; 
        
        min_x = min(min_x, x);
        min_y = min(min_y, y);

        planar.push_back({x, y});
    }

    for (auto &[x, y] : planar) {
        x = x - min_x;
        y = y - min_y;
    }
    return planar;
}


double heuristic(int start, int goal) {
    static vector<Node> planar = get_planar();

    start--; goal--;

    double delta_x = planar[start].x - planar[goal].x;
    double delta_y = planar[start].y - planar[goal].y;
    return sqrt( delta_x*delta_x + delta_y*delta_y );
}

string get_gedung(int gedung) {
    switch (gedung) {
        case 1: return "Agape";
        case 2: return "Biblos";
        case 3: return "Chara";
        case 4: return "Didaktos";
        case 5: return "Euodia";
        case 6: return "Filia";
        case 7: return "Gnosis";
        case 8: return "Hagios";
        case 9: return "Iama";
        case 10: return "Koinonia";
        case 11: return "Logos";
        case 12: return "Makarios";
        default: return "Unknown";
    }
}


void solve() {
    for (int i = 1; i <= n; i++) {
        cout << i << " = " << get_gedung(i) << '\n';
    }

    int start, goal;
    cout << "Asal gedung (1 - 12): ";
    cin >> start;
    cout << "Tujuan gedung (1 - 12): ";
    cin >> goal;

    if (start < 1 || start > n || goal < 1 || goal > n) {
        cout << "Input tidak valid.\n";
        return;
    }

    int m;
    cout << "Jumlah edge (jalur): ";
    cin >> m;

    vector<vector<pair<double, int>>> neighbours(n + 1);

    cout << "Masukkan jalur antar gedung (format: a b jarak):\n";
    for (int i = 0; i < m; i++) {
        int a, b; double w;
        cin >> a >> b >> w;
        neighbours[a].emplace_back(w, b);
        neighbours[b].emplace_back(w, a);
    }

    using T = pair<double, int>; // f_score, node
    priority_queue<T, vector<T>, greater<T>> pq;

    vector<double> g_score(n + 1, 1e18);
    vector<int> path(n + 1, -1);

    g_score[start] = 0;
    pq.emplace(heuristic(start, goal), start);

    while (!pq.empty()) {
        auto [f, current] = pq.top(); pq.pop();

        if (current == goal) {
            vector<int> ans;
            while (current != -1) {
                ans.push_back(current);
                current = path[current];
            }
            reverse(ans.begin(), ans.end());

            cout << "Rute terbaik:\n";
            for (int i = 0; i < (int)ans.size(); i++) {
                cout << get_gedung(ans[i]);
                if (i != (int)ans.size() - 1) cout << " -> ";
            }
            cout << "\nTotal jarak: " << g_score[goal] << '\n';
            return;
        }

        for (auto &[cost, neighbor] : neighbours[current]) {
            double tentative_g = g_score[current] + cost;
            if (tentative_g < g_score[neighbor]) {
                g_score[neighbor] = tentative_g;
                double f_score = tentative_g + heuristic(neighbor, goal);
                pq.emplace(f_score, neighbor);
                path[neighbor] = current;
            }
        }
    }

    cout << "Tidak ada jalur yang ditemukan.\n";
}



int main() {
    solve();
}

File: AStar.java
import java.util.*;

public class AStar {

    static final int N = 12; // jumlah gedung
    static final long INF = Long.MAX_VALUE;

    static class Pos {
        double lat, lon;
        Pos(double lat, double lon) { this.lat = lat; this.lon = lon; }
    }

    static Map<Integer, Pos> coordinates = new HashMap<>();
    static {
        coordinates.put(1, new Pos(-7.786892527493072, 110.37811638436851)); // Agape
        coordinates.put(2, new Pos(-7.786562258126049, 110.37825953130029)); // Biblos
        coordinates.put(3, new Pos(-7.78631777007192, 110.37842314604282));  // Chara
        coordinates.put(4, new Pos(-7.786012910895829, 110.37836850018982)); // Didaktos
        coordinates.put(5, new Pos(-7.785926844542299, 110.37811203519209)); // Euodia
        coordinates.put(6, new Pos(-7.786487300114521, 110.37804171414969)); // Filia
        coordinates.put(7, new Pos(-7.785753220073825, 110.37850078264714)); // Gnosis
        coordinates.put(8, new Pos(-7.785523171156505, 110.37836550377382)); // Hagios
        coordinates.put(9, new Pos(-7.7852138021969335, 110.37855178209473));// Iama
        coordinates.put(10, new Pos(-7.784999599965238, 110.37854751118998));// Koinonia
        coordinates.put(11, new Pos(-7.78528341456613, 110.37821452033853)); // Logos
        coordinates.put(12, new Pos(-7.785369062426726, 110.37854481744473));// Makarios
    }

    static String getGedung(int gedung) {
        switch (gedung) {
            case 1: return "Agape";
            case 2: return "Biblos";
            case 3: return "Chara";
            case 4: return "Didaktos";
            case 5: return "Euodia";
            case 6: return "Filia";
            case 7: return "Gnosis";
            case 8: return "Hagios";
            case 9: return "Iama";
            case 10: return "Koinonia";
            case 11: return "Logos";
            case 12: return "Makarios";
            default: return "Unknown";
        }
    }

    static double heuristic(int a, int b) {
        Pos pa = coordinates.get(a);
        Pos pb = coordinates.get(b);
        double dx = pb.lon - pa.lon;
        double dy = pb.lat - pa.lat;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public static void solve(Scanner sc) {

        for (int i = 1; i <= N; i++) {
            System.out.println(i + " = " + getGedung(i));
        }

        System.out.print("Asal gedung (1 - 12): ");
        int start = sc.nextInt();
        System.out.print("Tujuan gedung (1 - 12): ");
        int goal = sc.nextInt();

        if (start < 1 || start > N || goal < 1 || goal > N) {
            System.out.println("Input tidak valid.");
            return;
        }

        System.out.print("Jumlah edge (jalur): ");
        int m = sc.nextInt();

        List<List<int[]>> neighbours = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            neighbours.add(new ArrayList<>());
        }

        System.out.println("Masukkan jalur antar gedung (format: a b jarak):");
        for (int i = 0; i < m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            int w = sc.nextInt();
            neighbours.get(a).add(new int[]{w, b});
            neighbours.get(b).add(new int[]{w, a});
        }

        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));
        double[] g_score = new double[N + 1];
        Arrays.fill(g_score, Double.MAX_VALUE);
        g_score[start] = 0;

        int[] path = new int[N + 1];
        Arrays.fill(path, -1);

        pq.add(new double[]{heuristic(start, goal), start});

        while (!pq.isEmpty()) {
            double[] current = pq.poll();
            double f = current[0];
            int u = (int) current[1];

            if (u == goal) {
                List<Integer> ans = new ArrayList<>();
                while (u != -1) {
                    ans.add(u);
                    u = path[u];
                }
                Collections.reverse(ans);

                System.out.println("Rute terbaik:");
                for (int i = 0; i < ans.size(); i++) {
                    System.out.print(getGedung(ans.get(i)));
                    if (i != ans.size() - 1) System.out.print(" -> ");
                }
                return;
            }

            for (int[] edge : neighbours.get(u)) {
                int cost = edge[0];
                int v = edge[1];
                double tentative_g = g_score[u] + cost;
                if (tentative_g < g_score[v]) {
                    g_score[v] = tentative_g;
                    double f_score = tentative_g + heuristic(v, goal);
                    pq.add(new double[]{f_score, v});
                    path[v] = u;
                }
            }
        }

        System.out.println("Tidak ada jalur yang ditemukan.");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        solve(sc);
    }
}
